lib/src/main/java/com/diffplug/spotless/shell/ShfmtStep.java

```diff
@@ -86,25 +86,40 @@ static class State implements Serializable {
 		// used for up-to-date checks and caching
 		final String version;
 		final transient ForeignExe exe;
-		// used for executing
-		private transient @Nullable List<String> args;
 
 		State(ShfmtStep step, ForeignExe pathToExe) {
 			this.version = step.version;
 			this.exe = Objects.requireNonNull(pathToExe);
 		}
 
+		FormatterFunc.Closeable toFunc() {
+			ProcessRunner runner = new ProcessRunner();
+			return FormatterFunc.Closeable.ofDangerous(runner, new ShfmtFilePathPassingFormatterFunc(runner, exe));
+		}
+	}
+
+	private static class ShfmtFilePathPassingFormatterFunc implements FormatterFunc.NeedsFile {
+		// used for executing
+		private transient @Nullable List<String> args;
+		final transient ForeignExe exe;
+		final transient ProcessRunner runner;
+
+		ShfmtFilePathPassingFormatterFunc(ProcessRunner runner, ForeignExe exe) {
+			this.runner = runner;
+			this.exe = exe;
+		}
+
+		@Override
+		public String applyWithFile(String unix, File file) throws Exception {
+			return format(runner, unix, file);
+		}
+
 		String format(ProcessRunner runner, String input, File file) throws IOException, InterruptedException {
 			if (args == null) {
-				args = List.of(exe.confirmVersionAndGetAbsolutePath(), "-i", "2", "-ci");
+				args = List.of(exe.confirmVersionAndGetAbsolutePath(), file.getAbsolutePath());
```

Comment:
What if the second run, still using the first file path?

Reply:
Hm.. Looks like this formatter doesn't work when there is more than one file to format. The second file ends up containing the formatted contents of the first file that was formatted. Do you have any suggestions here? I assumed that the `file` variable was going to contain the current file that needs formatted, but that doesn't seem to actually be the case.

Reply:
@Goooler this should all be fixed now. The solution turned out to be simpler than I thought in the end.

Reply:
From my observation, `--filename` is not a necessary param for shfmt, we can remove it. Am I missing anything here?

Reply:
I've tried to remove it, seems tests passed, see

https://github.com/Goooler/spotless/commit/62fcf4e294f995d69214ab18c2fdabb168ddcf18
https://github.com/Goooler/spotless/actions/runs/7753448370

Reply:
The CI tests will pass because the tests aren't running for shfmt there. That's something that should probably be added though. See Ned's comment here: https://github.com/diffplug/spotless/pull/1998#discussion_r1454632452. I can work on adding CI support in this PR. I think that would be beneficial here. I was thinking of doing it in a subsequent PR, but now might be the best time to do it.

Reply:
If we must input the file path here, it might be something like https://github.com/Goooler/spotless/commit/ae75240a71a056b791df0a940acf2ab0f6e0ddc8?

Reply:
> If we must input the file path here, it might be something like [Goooler@ae75240](https://github.com/Goooler/spotless/commit/ae75240a71a056b791df0a940acf2ab0f6e0ddc8)?

Would you be fine with something like the following instead? I don't think `args` needs to be an instance variable as of right now. If args need provided to the `State` constructor eventually, then someone can refactor it for their needs. Unless `args` is somehow used implicitly for something that I'm unaware of.
```java
String format(ProcessRunner runner, String input, File file) throws IOException, InterruptedException {
	final List<String> args = List.of(exe.confirmVersionAndGetAbsolutePath(), "--filename", file.getPath());
	return runner.exec(input.getBytes(StandardCharsets.UTF_8), args).assertExitZero(StandardCharsets.UTF_8);
}
```

Reply:
The only concern is if it's necessary to check the executable file for every formatting. Otherwise, I'm okay with your snippet.